File: v1/LLD/Design-Patterns/Facade/code/Facade.cpp
// Facade Pattern — OrderFacade coordinating several subsystems behind one call.
// Compile: g++ -std=c++14 Facade.cpp -o facade
// Run:     ./facade

#include <iostream>
#include <string>

class InventoryService {
public:
    bool reserveItem(const std::string& item) {
        std::cout << "Inventory: reserved '" << item << "'" << std::endl;
        return true;
    }
};

class PaymentService {
public:
    bool charge(double amount) {
        std::cout << "Payment: charged $" << amount << std::endl;
        return true;
    }
};

class ShippingService {
public:
    void scheduleShipment(const std::string& item) {
        std::cout << "Shipping: scheduled shipment for '" << item << "'" << std::endl;
    }
};

class NotificationService {
public:
    void sendConfirmation(const std::string& item) {
        std::cout << "Notification: confirmation email sent for '" << item << "'" << std::endl;
    }
};

class OrderFacade {
public:
    void placeOrder(const std::string& item, double price) {
        std::cout << "--- Placing order for " << item << " ---" << std::endl;
        if (inventory_.reserveItem(item) && payment_.charge(price)) {
            shipping_.scheduleShipment(item);
            notification_.sendConfirmation(item);
            std::cout << "Order complete." << std::endl;
        }
    }

private:
    InventoryService inventory_;
    PaymentService payment_;
    ShippingService shipping_;
    NotificationService notification_;
};

int main() {
    OrderFacade orderFacade;
    orderFacade.placeOrder("Wireless Mouse", 24.99);
    return 0;
}

File: v1/LLD/Design-Patterns/Facade/code/Facade.java
// Facade Pattern — OrderFacade coordinating several subsystems behind one call.
// Compile: javac Facade.java
// Run:     java Facade

class InventoryService {
    boolean reserveItem(String item) {
        System.out.println("Inventory: reserved '" + item + "'");
        return true;
    }
}

class PaymentService {
    boolean charge(double amount) {
        System.out.println("Payment: charged $" + amount);
        return true;
    }
}

class ShippingService {
    void scheduleShipment(String item) {
        System.out.println("Shipping: scheduled shipment for '" + item + "'");
    }
}

class NotificationService {
    void sendConfirmation(String item) {
        System.out.println("Notification: confirmation email sent for '" + item + "'");
    }
}

// The Facade: one simple method hiding the coordination of four subsystems.
class OrderFacade {
    private InventoryService inventory = new InventoryService();
    private PaymentService payment = new PaymentService();
    private ShippingService shipping = new ShippingService();
    private NotificationService notification = new NotificationService();

    void placeOrder(String item, double price) {
        System.out.println("--- Placing order for " + item + " ---");
        if (inventory.reserveItem(item) && payment.charge(price)) {
            shipping.scheduleShipment(item);
            notification.sendConfirmation(item);
            System.out.println("Order complete.");
        }
    }
}

public class Facade {
    public static void main(String[] args) {
        OrderFacade orderFacade = new OrderFacade();
        orderFacade.placeOrder("Wireless Mouse", 24.99);
    }
}
